Add tikv-br doc (#325)
tikv/website: content/docs/dev/concepts/explore-tikv-features/backup-restore.md

## RawKV Backup & Restore

**[TiKV Backup & Restore (TiKV-BR)](https://github.com/tikv/migration/tree/main/br)** is a command-line tool for distributed backup and restoration of the TiKV cluster data.

## Architecture

{{< figure
    src="/img/docs/tikv-br.png"
    caption="TiKV-BR architecture"
    number="1" >}}

### Recommended Deployment Configuration
- In production environments, deploy `TiKV-BR` on a node with at least 8 cores CPU and 16 GB memory. Select an appropriate OS version by following [Linux OS version requirements](https://docs.pingcap.com/tidb/dev/hardware-and-software-requirements#linux-os-version-requirements).

> - Save backup data to Amazon S3 or one mounted network disk on all `TiKV-BR` and `TiKV` nodes.

Reviewer: ```suggestion
- Save backup data to Amazon S3 or other shared storage, for example mounting a NFS on all `TiKV-BR` and `TiKV` nodes.
```

- Allocate sufficient resources for backup and restoration.

- TiKV-BR only support raw data backup/restoration in TiKV cluster with version >= `5.0.0`.

> TiKV-BR, TiKV nodes, and the backup storage system should provide network bandwidth that is greater than the backup speed. If the target cluster is particularly large, the threshold of backup and restoration speed is limited by the bandwidth of the backup network.  

Reviewer: ```suggestion
TiKV-BR, TiKV nodes, and the backup storage system should provide network bandwidth that is greater than the backup speed. If the target cluster is particularly large, the backup and restoration speed highly depends on the bandwidth of storage system bandwidth.  
```

Author: Addressed and apply some suggestions from google docs, PTAL, thanks.

A progress bar is displayed in the terminal during the backup. When the progress bar advances to 100%, the backup is complete. The progress bar is displayed as follows:
```
br backup raw \
    --pd "${PDIP}:2379" \
    --storage "s3://backup-data/2022-09-16/" \
    --log-file backupfull.log
Backup Raw <---------/................................................> 17.12%.
```

After backup finish, the result message is displayed as follows:
```
[2022/09/20 18:01:10.125 +08:00] [INFO] [collector.go:67] ["Raw backup success summary"] [total-ranges=3] [ranges-succeed=3] [ranges-failed=0] [backup-total-regions=3] [total-take=5.050265883s] [backup-ts=436120585518448641] [total-kv=100000] [total-kv-size=108.7MB] [average-speed=21.11MB/s] [backup-data-size(after-compressed)=78.3MB]
```
Explanations for the above message are as follows: 
- `total-ranges`: Number of ranges that the whole backup task is split to. Equals to `ranges-succeed` + `ranges-failed`.
- `ranges-succeed`: Number of succeeded ranges.
- `ranges-failed`: Number of failed ranges.
- `backup-total-regions`: The tikv regions that backup takes.
- `total-take`: The backup duration.

> - `backup-ts`: The backup start timestamp, only take effect for APIV2 TiKV cluster, which can be used as `start-ts` of `TiKV-CDC` when creating replication tasks. Refer to [Create a replication task](https://github.com/tikv/migration/blob/main/cdc/README.md#create-a-replication-task).

Reviewer 2: ```suggestion
- `backup-ts`: The backup start timestamp, only take effect for API V2 TiKV cluster, which can be used as `start-ts` of `TiKV-CDC` when creating replication tasks. Refer to [Create a replication task](https://github.com/tikv/migration/blob/main/cdc/README.md#create-a-replication-task).
```

> - `total-kv`: Total kv count in backup files.

Reviewer 2: ```suggestion
- `total-kv`: Total number of key-value pairs in backup files.
```

> - `total-kv-size`: Total kv size in backup files. Note that this is the original size before compression.

Reviewer 2: ```suggestion
- `total-kv-size`: Total size of key-value pairs in backup files. Note that this is the original size before compression.
```

- `average-speed`: The backup speed, which approximately equals to `total-kv-size` / `total-take`.
- `backup-data-size(after-compressed)`: The backup file size.

#### Restore Raw Data

To restore raw data to the cluster, execute the `tikv-br restore raw` command. To get help on this command, execute `tikv-br restore raw -h` or `tikv-br restore raw --help`.
For example, restore the raw backup files in s3 `/backup-data/2022-09-16` to `TiKV` cluster.

```
export AWS_ACCESS_KEY_ID=&{AWS_KEY_ID};
export AWS_SECRET_ACCESS_KEY=&{AWS_KEY};
tikv-br restore raw \
    --pd "${PDIP}:2379" \
    --storage "s3://backup-data/2022-09-16/" \
    --ratelimit 128 \
    --log-file restoreraw.log
```
Explanations for some options in the above command are as follows:

- `--ratelimit`: The maximum speed at which a restoration operation is performed (MiB/s) on each `TiKV` node.

> - `--log-file`: Writing the TiKV-BR log to the `restorefull.log` file.

Reviewer 2: ```suggestion
- `--log-file`: Writing the TiKV-BR log to the `restoreraw.log` file.
```

A progress bar is displayed in the terminal during the restoration. When the progress bar advances to 100%, the restoration is complete. The progress bar is displayed as follows:
```
tikv-br restore raw \
    --pd "${PDIP}:2379" \
    --storage "s3://backup-data/2022-09-16/" \
    --ratelimit 128 \
    --log-file restoreraw.log
Restore Raw <---------/...............................................> 17.12%.
```

After restoration finish, the result message is displayed as follows:
```
[2022/09/20 18:02:12.540 +08:00] [INFO] [collector.go:67] ["Raw restore success summary"] [total-ranges=3] [ranges-succeed=3] [ranges-failed=0] [restore-files=3] [total-take=950.460846ms] [restore-data-size(after-compressed)=78.3MB] [total-kv=100000] [total-kv-size=108.7MB] [average-speed=114.4MB/s]
```
Explanations for the above message are as follows: 
- `total-ranges`: Number of ranges that the whole backup task is split to. Equals to `ranges-succeed` + `ranges-failed`.

> - `ranges-succeed`: Number of succeeded ranges.

Reviewer 2: ```suggestion
- `ranges-succeed`: Number of successful ranges.
```

- `ranges-failed`: Number of failed ranges.
- `restore-files`: Number of restored files.
- `total-take`: The restoration duration.

> - `total-kv`: Total restored kv count.

Reviewer 2: ```suggestion
- `total-kv`: Total number of restored key-value pairs.
```

> - `total-kv-size`: Total restored kv size. Note that this is the original size before compression.

Reviewer 2: ```suggestion
- `total-kv-size`: Total size restored key-value pairs. Note that this is the original size before compression.
```

- `average-speed`: The restoration speed, which approximately equals to `total-kv-size` / `total-take`.
- `restore-data-size(after-compressed)`: The restoration file size.

### Data Verification of Backup & Restore

> TiKV-BR can do checksum between TiKV cluster and backup files after backup or restoration finish with the config `--checksum=true`. Checksum is using the [checksum](https://github.com/tikv/client-go/blob/ffaaf7131a8df6ab4e858bf27e39cd7445cf7929/rawkv/rawkv.go#L584) interface in TiKV [client-go](https://github.com/tikv/client-go), which send checksum request to all TiKV regions to calculate the checksum of all **VALID** data. Then compare to the checksum value of backup files which is calculated during backup process.

Reviewer 2: TODO: change link of `checksum` after we add a `checksum` introduction [here](https://deploy-preview-325--tikv.netlify.app/docs/dev/develop/rawkv/introduction/).

> In some scenario, data is stored in TiKV with [TTL](https://docs.pingcap.com/tidb/stable/tikv-configuration-file#enable-ttl). If data is expired during backup & restore, the persisted checksum in backup files is different from the checksum of TiKV cluster. So checksum should not enabled in this scenario. User can perform a full comparison for all existing non-expired data between backup cluster and restore cluster with [scan](https://github.com/tikv/client-go/blob/ffaaf7131a8df6ab4e858bf27e39cd7445cf7929/rawkv/rawkv.go#L492) interface in [client-go](https://github.com/tikv/client-go).

Reviewer 2: ```suggestion
Please note that if data is stored in TiKV with [TTL](./ttl), and expiration happens during backup or restore, the persisted checksum in backup files must be different from the checksum of TiKV cluster. So checksum should **NOT** be enabled in this scenario. To verify correctness of backup and restoration in this scenario, you can perform a full comparison for all existing non-expired data between backup cluster and restore cluster by using [scan](../../develop/rawkv/scan) interface.
```
